docs/source/design/dataset.md

```diff
@@ -0,0 +1,340 @@
+Olive Dataset Design
+====================
+
+### Overview
+
+Currently, we have a few different dataset implementations in Olive. We should unify them under standardized Olive.Dataset/Olive.Dataloader interface which will be introduced in this design document.
+
```

Comment:
We should update this design doc to reflect the new implementation. Once this PR is merged into the main this design document should not discuss what will be introduced/changed by this PR.

Reply:
For sure, will update it.

Reply:
Update.